compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -1278,7 +1288,7 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
 
         let tilde_const_allowed =
             matches!(fk.header(), Some(FnHeader { constness: ast::Const::Yes(_), .. }))
-                || matches!(fk.ctxt(), Some(FnCtxt::Assoc(_)));
+                || (matches!(fk.ctxt(), Some(FnCtxt::Assoc(_))) && self.in_const_trait);
```

Comment:
would this be more readable?

```suggestion
                || matches!(fk.ctxt(), Some(FnCtxt::Assoc(_)) if self.in_const_trait);
```